File: app/src/main/java/org/worshipsongs/activity/NavigationDrawerActivity.java
package org.worshipsongs.activity;

import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.content.res.ColorStateList;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.preference.PreferenceManager;
import android.support.v7.widget.Toolbar;
import android.util.TypedValue;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import org.worshipsongs.CommonConstants;
import org.worshipsongs.R;
import org.worshipsongs.fragment.HomeFragment;
import org.worshipsongs.registry.FragmentRegistry;
import org.worshipsongs.service.PresentationScreenService;
import org.worshipsongs.service.SongService;
import org.worshipsongs.utils.CommonUtils;
import org.worshipsongs.utils.ThemeUtils;


public class NavigationDrawerActivity extends AbstractAppCompactActivity implements NavigationView.OnNavigationItemSelectedListener
{
    private static final int UPDATE_DB_REQUEST_CODE = 555;
    private static final String SENDER_MAIL = "appfeedback@mcruncher.com";
    private SharedPreferences sharedPreferences;
    private PresentationScreenService presentationScreenService;

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        ThemeUtils.setNoActionBarTheme(this);
        setContentView(R.layout.activity_main);
        presentationScreenService = new PresentationScreenService(this);
        setSongCount();
        setDrawerLayout();
        setNavigationView(savedInstanceState);
    }

    private void setSongCount()
    {
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);
        if (!sharedPreferences.getAll().containsKey(CommonConstants.NO_OF_SONGS)) {
            SongService songService = new SongService(this);
            long count = songService.count();
            sharedPreferences.edit().putLong(CommonConstants.NO_OF_SONGS, count).apply();
        }
    }

    private void setDrawerLayout()
    {
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        if (Build.VERSION.SDK_INT == Build.VERSION_CODES.LOLLIPOP) {
            toolbar.setElevation(0);
        }
        setSupportActionBar(toolbar);

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(this, drawer, toolbar,
                R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.addDrawerListener(toggle);
        toggle.syncState();
    }

    private void setNavigationView(Bundle savedInstanceState)
    {
        NavigationView navigationView = (NavigationView) findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);
        setHeaderView(navigationView);
        if (savedInstanceState == null) {
            MenuItem item = navigationView.getMenu().getItem(0);
            onNavigationItemSelected(item);
            navigationView.setCheckedItem(R.id.home);
        }
        ColorStateList colorStateList = getColorStateList();
        navigationView.setItemTextColor(colorStateList);
        navigationView.setItemIconTintList(colorStateList);
        TextView versionTextView = (TextView) navigationView.findViewById(R.id.version);
        versionTextView.setText(getString(R.string.version, CommonUtils.getProjectVersion()));
    }

    private void setHeaderView(NavigationView navigationView)
    {
        View headerView = navigationView.getHeaderView(0);
        TextView headerSubTitleTextView = (TextView) headerView.findViewById(R.id.header_subtitle);
        headerSubTitleTextView.setText(getString(R.string.noOfSongsAvailable,
                sharedPreferences.getLong(CommonConstants.NO_OF_SONGS, 0)));
    }


    @NonNull
    private ColorStateList getColorStateList()
    {
        TypedValue typedValue = new TypedValue();
        getTheme().resolveAttribute(android.R.attr.textColor, typedValue, true);
        int[][] state = new int[][]{
                new int[]{-android.R.attr.state_enabled}, // disabled
                new int[]{android.R.attr.state_enabled}, // enabled
                new int[]{-android.R.attr.state_checked}, // unchecked
                new int[]{android.R.attr.state_pressed}  // pressed
        };
        int[] color = new int[]{
                typedValue.data,
                typedValue.data,
                typedValue.data,
                typedValue.data
        };
        return new ColorStateList(state, color);
    }

    @Override
    public void onBackPressed()
    {
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
            finish();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu)
    {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item)
    {
        return super.onOptionsItemSelected(item);
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(MenuItem item)
    {
        switch (item.getItemId()) {
            case R.id.home:
                setHomeView();
                break;
            case R.id.updateSongs:
                setUpdateView();
                break;
            case R.id.settings:
                startActivity(new Intent(NavigationDrawerActivity.this,
                        UserSettingActivity.class));
                break;
            case R.id.rateUs:
                setRateUsView();
                break;
            case R.id.share:
                setShareView();
                break;
            case R.id.feedback:
                setEmail();
                break;
        }
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }

    private void setHomeView()
    {
        FragmentManager fragmentManager = getSupportFragmentManager();
        HomeFragment existingHomeTabFragment = (HomeFragment) fragmentManager
                .findFragmentByTag(HomeFragment.class.getSimpleName());
        if (existingHomeTabFragment == null) {
            Fragment fragment = HomeFragment.newInstance();
            fragment.setArguments(getIntent().getExtras());
            FragmentTransaction transaction = fragmentManager.beginTransaction();
            transaction.replace(R.id.content_frame, fragment, HomeFragment.class.getSimpleName());
            transaction.addToBackStack(null);
            transaction.commit();
        }
    }


    private void setUpdateView()
    {
        Intent updateSongs = new Intent(NavigationDrawerActivity.this,
                UpdateSongsDatabaseActivity.class);
        startActivityForResult(updateSongs, UPDATE_DB_REQUEST_CODE);
    }

    private void setRateUsView()
    {
        Uri uri = Uri.parse("market://details?id=" + NavigationDrawerActivity.
                this.getApplicationContext().getPackageName());
        Intent goToMarket = new Intent(Intent.ACTION_VIEW, uri);
        goToMarket.addFlags(getFlags());
        try {
            startActivity(goToMarket);
        } catch (ActivityNotFoundException e) {
            startActivity(new Intent(Intent.ACTION_VIEW,
                    Uri.parse("http://play.google.com/store/apps/details?id=" +
                            NavigationDrawerActivity.this.getApplicationContext().getPackageName())));
        }
    }

    private void setShareView()
    {
        Intent shareIntent = new Intent(Intent.ACTION_SEND);
        shareIntent.putExtra(Intent.EXTRA_TEXT,
                getString(R.string.app_description) + getString(R.string.app_download_info));
        shareIntent.setType("text/plain");
        Intent intent = Intent.createChooser(shareIntent, getString(R.string.share) + " "
                + getString(R.string.app_name) + " in");
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(shareIntent);
    }

    int getFlags()
    {
        return (Intent.FLAG_ACTIVITY_NO_HISTORY |
                Intent.FLAG_ACTIVITY_NEW_DOCUMENT |
                Intent.FLAG_ACTIVITY_MULTIPLE_TASK);
    }

    private void setEmail()
    {
        Intent mailIntent = new Intent(Intent.ACTION_SENDTO);
        mailIntent.setData(Uri.parse("mailto:" + SENDER_MAIL));
        mailIntent.putExtra(Intent.EXTRA_EMAIL, "");
        mailIntent.putExtra(Intent.EXTRA_SUBJECT, getEmailSubject(getApplicationContext()));
        startActivity(Intent.createChooser(mailIntent, ""));
    }

    String getEmailSubject(Context context)
    {
        try {
            String versionName = context.getPackageManager().getPackageInfo(context.getPackageName()
                    , 0).versionName;
            return String.format(context.getString(R.string.feedback_subject), versionName);
        } catch (PackageManager.NameNotFoundException e) {
            return getString(R.string.feedback);
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data)
    {
        switch (requestCode) {
            case UPDATE_DB_REQUEST_CODE:
                long noOfSongs = sharedPreferences.getLong(CommonConstants.NO_OF_SONGS, 0);
                sharedPreferences.edit().putLong(CommonConstants.NO_OF_SONGS, noOfSongs).apply();
                TextView headerSubTitleTextView = (TextView) findViewById(R.id.header_subtitle);
                headerSubTitleTextView.setText(getString(R.string.noOfSongsAvailable, noOfSongs));
                break;
            default:
                break;
        }
    }

    @Override
    protected void onPause()
    {
        super.onPause();
        presentationScreenService.onPause();
    }

    @Override
    protected void onStop()
    {
        super.onStop();
        presentationScreenService.onResume();
    }
}

File: app/src/main/java/org/worshipsongs/activity/NavigationDrawerActivity.kt
package org.worshipsongs.activity

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.content.res.ColorStateList
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.util.Log

import androidx.core.view.GravityCompat


import android.util.TypedValue
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.widget.Toolbar
import androidx.drawerlayout.widget.DrawerLayout
import androidx.preference.PreferenceManager
import com.google.android.material.navigation.NavigationView
import org.worshipsongs.CommonConstants
import org.worshipsongs.R
import org.worshipsongs.fragment.HomeFragment
import org.worshipsongs.service.PresentationScreenService
import org.worshipsongs.service.SongService
import org.worshipsongs.utils.CommonUtils
import org.worshipsongs.utils.ThemeUtils


class NavigationDrawerActivity : AbstractAppCompactActivity(), NavigationView.OnNavigationItemSelectedListener {
    private var sharedPreferences: SharedPreferences? = null
    private var presentationScreenService: PresentationScreenService? = null

    val colorStateList: ColorStateList
        get() {
            val typedValue = TypedValue()
            theme.resolveAttribute(android.R.attr.textColor, typedValue, true)
            val state = arrayOf(intArrayOf(-android.R.attr.state_enabled), intArrayOf(android.R.attr.state_enabled), intArrayOf(-android.R.attr.state_checked), intArrayOf(android.R.attr.state_pressed))
            val color = intArrayOf(typedValue.data, typedValue.data, typedValue.data, typedValue.data)
            return ColorStateList(state, color)
        }

    val flags: Int
        get() = Intent.FLAG_ACTIVITY_NO_HISTORY or
                Intent.FLAG_ACTIVITY_NEW_DOCUMENT or
                Intent.FLAG_ACTIVITY_MULTIPLE_TASK

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ThemeUtils.setNoActionBarTheme(this)
        setContentView(R.layout.activity_main)
        presentationScreenService = PresentationScreenService(this)
        setSongCount()
        setDrawerLayout()
        setNavigationView(savedInstanceState)
    }

    private fun setSongCount() {
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this)
        if (!sharedPreferences!!.all.containsKey(CommonConstants.NO_OF_SONGS)) {
            val songService = SongService(this)
            val count = songService.count()
            sharedPreferences!!.edit().putLong(CommonConstants.NO_OF_SONGS, count).apply()
        }
    }

    private fun setDrawerLayout() {
        val toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        if (Build.VERSION.SDK_INT == Build.VERSION_CODES.LOLLIPOP) {
            toolbar.elevation = 0f
        }
        setSupportActionBar(toolbar)

        val drawer = findViewById<View>(R.id.drawer_layout) as DrawerLayout
        val toggle = ActionBarDrawerToggle(this, drawer, toolbar,
                R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        drawer.addDrawerListener(toggle)
        toggle.syncState()
    }

    private fun setNavigationView(savedInstanceState: Bundle?) {
        val navigationView = findViewById<View>(R.id.nav_view) as NavigationView
        navigationView.setNavigationItemSelectedListener(this)
        setHeaderView(navigationView)
        if (savedInstanceState == null) {
            val item = navigationView.menu.getItem(0)
            onNavigationItemSelected(item)
            navigationView.setCheckedItem(R.id.home)
        }
        val colorStateList = colorStateList
        navigationView.itemTextColor = colorStateList
        navigationView.itemIconTintList = colorStateList
        val versionTextView = navigationView.findViewById<View>(R.id.version) as TextView
        versionTextView.text = getString(R.string.version, CommonUtils.projectVersion)
    }

    private fun setHeaderView(navigationView: NavigationView) {
        val headerView = navigationView.getHeaderView(0)
        val headerSubTitleTextView = headerView.findViewById<View>(R.id.header_subtitle) as TextView
        headerSubTitleTextView.text = getString(R.string.noOfSongsAvailable,
                sharedPreferences!!.getLong(CommonConstants.NO_OF_SONGS, 0))
    }

    override fun onBackPressed() {
        val drawer = findViewById<View>(R.id.drawer_layout) as DrawerLayout
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
            finish()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return super.onOptionsItemSelected(item)
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.home -> setHomeView()
            R.id.updateSongs -> setUpdateView()
            R.id.settings -> startActivity(Intent(this@NavigationDrawerActivity,
                    UserSettingActivity::class.java))
            R.id.rateUs -> setRateUsView()
            R.id.share -> setShareView()
            R.id.feedback -> setEmail()
        }
        val drawer = findViewById<View>(R.id.drawer_layout) as DrawerLayout
        drawer.closeDrawer(GravityCompat.START)
        return true
    }

    private fun setHomeView() {
        val fragmentManager = supportFragmentManager
        val existingHomeTabFragment = fragmentManager
                .findFragmentByTag(HomeFragment::class.java!!.getSimpleName()) as HomeFragment?
        if (existingHomeTabFragment == null) {
            val fragment = HomeFragment.newInstance()
            fragment.arguments = intent.extras
            val transaction = fragmentManager.beginTransaction()
            transaction.replace(R.id.content_frame, fragment, HomeFragment::class.java!!.getSimpleName())
            transaction.addToBackStack(null)
            transaction.commit()
        }
    }


    private fun setUpdateView() {
        val updateSongs = Intent(this@NavigationDrawerActivity,
                UpdateSongsDatabaseActivity::class.java)
        startActivityForResult(updateSongs, UPDATE_DB_REQUEST_CODE)
    }

    private fun setRateUsView() {
        val uri = Uri.parse("market://details?id=" + this@NavigationDrawerActivity.applicationContext.packageName)
        val goToMarket = Intent(Intent.ACTION_VIEW, uri)
        goToMarket.addFlags(flags)
        try {
            startActivity(goToMarket)
        } catch (e: ActivityNotFoundException) {
            startActivity(Intent(Intent.ACTION_VIEW,
                    Uri.parse("http://play.google.com/store/apps/details?id=" + this@NavigationDrawerActivity.applicationContext.packageName)))
        }

    }

    private fun setShareView() {
        val shareIntent = Intent(Intent.ACTION_SEND)
        shareIntent.putExtra(Intent.EXTRA_TEXT,
                getString(R.string.app_description) + getString(R.string.app_download_info))
        shareIntent.type = "text/plain"
        val intent = Intent.createChooser(shareIntent, getString(R.string.share) + " "
                + getString(R.string.app_name) + " in")
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        startActivity(shareIntent)
    }

    private fun setEmail() {
        val mailIntent = Intent(Intent.ACTION_SENDTO)
        mailIntent.data = Uri.parse("mailto:$SENDER_MAIL")
        mailIntent.putExtra(Intent.EXTRA_EMAIL, "")
        mailIntent.putExtra(Intent.EXTRA_SUBJECT, getEmailSubject(applicationContext))
        startActivity(Intent.createChooser(mailIntent, ""))
    }

     fun getEmailSubject(context: Context): String {
        try {
            val versionName = context.packageManager.getPackageInfo(context.packageName, 0).versionName
            return String.format(context.getString(R.string.feedback_subject), versionName)
        } catch (e: PackageManager.NameNotFoundException) {
            return getString(R.string.feedback)
        }

    }


    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            UPDATE_DB_REQUEST_CODE -> {
                val noOfSongs = sharedPreferences!!.getLong(CommonConstants.NO_OF_SONGS, 0)
                sharedPreferences!!.edit().putLong(CommonConstants.NO_OF_SONGS, noOfSongs).apply()
                val headerSubTitleTextView = findViewById<View>(R.id.header_subtitle) as TextView
                headerSubTitleTextView.text = getString(R.string.noOfSongsAvailable, noOfSongs)
            }
            else -> {
            }
        }
    }

    public override fun onResume()
    {
        super.onResume()
        presentationScreenService!!.onResume()
    }
    override fun onPause() {
        super.onPause()
        presentationScreenService!!.onPause()
    }

    override fun onStop() {
        super.onStop()
        presentationScreenService!!.onStop()
    }

    companion object {
        private val UPDATE_DB_REQUEST_CODE = 555
        private val SENDER_MAIL = "appfeedback@mcruncher.com"
    }
}
